File: Grind75/58-ConstructBSTPreorderInorder/cpp/ConstructBSTPreorderInorder.cpp
#include "ConstructBSTPreorderInorder.h"
#include <Tree.h>
#include <unordered_map>
#include <vector>

using namespace std;
using namespace BST;

TreeNode *build_recurse(
    const vector<int> &preorder, int index,
    const vector<int> &inorder,  int left, int right,
    const unordered_map<int, size_t> &iMap) {
  if (left > right || size_t(index) == preorder.size())
    return nullptr;
  TreeNode *root = new TreeNode(preorder[index]);
  int rootLoc = iMap.at(preorder[index]);
  root->left
    = build_recurse(preorder, index+1, inorder, left, rootLoc-1, iMap);
  root->right
    = build_recurse(preorder, index+1+rootLoc-left, inorder, rootLoc+1, right, iMap);
  return root;
}

TreeNode* Solution::buildTree(vector<int>& preorder, vector<int>& inorder) {
  unordered_map<int, size_t> iMap;
  for (size_t i = 0; i < inorder.size(); i++)
    iMap[inorder[i]] = i;
  return build_recurse(preorder, 0, inorder, 0, inorder.size()-1, iMap);
}

File: Grind75/58-ConstructBSTPreorderInorder/python/ConstructBSTPreorderInorder.py
#!/bin/python3

from __future__ import annotations
import sys
from typing import Dict
sys.path.append('../../Utility/python')
from PrettyPrint import *
from Colors import *
from Tree import *

class Solution:
  def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
    if len(preorder) == 0:
      return None
    iMap: Dict[int,int] = { v:i for i,v in enumerate(inorder) }
    def recurse(po: int, iol: int, ior) -> Optional[TreeNode]:
      if po >= len(preorder) or iol > ior:
        return None
      root = TreeNode(preorder[po])
      rootLoc = iMap[root.val]
      root.left  = recurse(po + 1, iol, rootLoc-1)
      root.right = recurse(po + (rootLoc-iol) + 1, rootLoc+1, ior)
      return root
    return recurse(0, 0, len(inorder)-1)

if __name__ == '__main__':
  soln = Solution()
  printTree(soln.buildTree([3,9,20,15,7],[9,3,15,20,7]))

